src/main/java/org/dataloader/DataLoaderHelper.java

```diff
@@ -546,4 +615,188 @@ private CompletableFuture<List<V>> setToValueCache(List<V> assembledValues, List
     private static <T> DispatchResult<T> emptyDispatchResult() {
         return (DispatchResult<T>) EMPTY_DISPATCH_RESULT;
     }
+
+    private class BatchObserverImpl implements BatchObserver<V> {
```

Comment:
The class is already a MONSTER - this should be there own package level classes.

We should move them to their own class - package protected say

Reply:
except you use inner methods so..... maybe they need to stay...